### PR TITLE
Sponsorship refactor

### LINKED ISSUE #2926: Rework removeEntryWithPossibleSponsorship to Create a Child LedgerTxn
The same will have to be done for related functions like `removeSignerWithPossibleSponsorship`. Overall, this should make the interface safer and easier to use.

### PR DESCRIPTION
**Pull request by jonjove**
# Description
Related to #3099
Resolves #2926

The primary goal of this PR is to simplify the maintenance of SponsorshipUtils.{h,cpp}. The way I originally wrote that code was too complex with multiple diverging and converging code paths. This version introduces a hierarchy of `Sponsorable` objects which enable a much simpler implementation by

1. allowing `LedgerTxnEntry` to be loaded repeatedly: these objects are never kept in memory, simplifying cases where multiple "roles" (eg. self and owner) are played by the same entry
2. making it clear what operations apply to different types: the public interface does not expose a way to remove sponsorship from a claimable balance
3. performing all operations within a nested `LedgerTxn`
4. using inheritance to share code in a streamlined manner

This is a draft because

- I haven't audited that it covers every case which the old implementation covered
- I haven't switched to the new implementation everywhere: there is an idiom where we load all offers that conflicts with (3) above

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

**Review by sisuresh — COMMENTED** [review]
I like this change. Much easier to follow what's actually going on with sponsorships.

I know this is still a draft and you're probably already aware, but I noticed a couple things and had one general comment - This PR creates a nested `LedgerTxn` in the sponsorship utils, so we should be careful about things like dangling references (thinking about `RevokeSponsorshipOpFrame` here).

File: src/transactions/RevokeSponsorshipOpFrame.cpp [hunk 1]
```diff
@@ -139,64 +142,35 @@ RevokeSponsorshipOpFrame::updateLedgerEntrySponsorship(AbstractLedgerTxn& ltx)
     if (wasEntrySponsored && willEntryBeSponsored)
     {
         // Transfer sponsorship
-        auto oldSponsoringAcc = loadAccount(ltx, *le.ext.v1().sponsoringID);
         auto const& se = sponsorship.currentGeneralized().sponsorshipEntry();
-        auto newSponsoringAcc = loadAccount(ltx, se.sponsoringID);
-        auto res = canTransferEntrySponsorship(header.current(), le,
-                                               oldSponsoringAcc.current(),
-                                               newSponsoringAcc.current());
+        auto sponsoringID = se.sponsoringID; // Need a copy
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/transactions/RevokeSponsorshipOpFrame.cpp:146]: Fixed.

File: src/transactions/ClawbackClaimableBalanceOpFrame.cpp [hunk 2]
```diff
@@ -62,12 +64,9 @@ ClawbackClaimableBalanceOpFrame::doApply(AbstractLedgerTxn& ltx)
         return false;
     }
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/transactions/ClawbackClaimableBalanceOpFrame.cpp:66]: Done, thanks.

**Comment by MonsieurNicolas** [comment]
closing: stale and unlikely to land anytime soon

**Pull request closed without merging**